### ISSUE TITLE
DirectoryInfo("H:\\").GetDirectories() throws FileNotFoundException when directory is the root and it is empty

### ISSUE THREAD

**Issue by stukselbax** [dotnet/runtime#33475]
Given `H:\` is the empty drive with no file or directory in it, and trying to call `GetDirectories()`:

```c#
namespace EmptyVolume
{
    using System;
    using System.IO;
    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                new DirectoryInfo("H:\\").GetDirectories();
                Console.WriteLine("Success!");
            }
            catch (FileNotFoundException e)
            {
                Console.WriteLine("Fail: {0}", e.ToString());
            }
        }
    }
}
```

throws:
```log
Fail: System.IO.FileNotFoundException: Could not find file 'H:\'.
File name: 'H:\'
   at System.IO.Enumeration.FileSystemEnumerator`1.GetData()
   at System.IO.Enumeration.FileSystemEnumerator`1.FindNextEntry()
   at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.IO.DirectoryInfo.GetDirectories(String searchPattern, EnumerationOptions enumerationOptions)
   at System.IO.DirectoryInfo.GetDirectories()
   at EmptyVolume.Program.Main(String[] args)
```

Tried on _netcoerapp2.1 (2.1.15), netcoerapp 3.1 (3.1.1)_. Also, if network share root volume used to `GetDirectories()`, and it is empty, e.g. `new DirectoryInfo("\\\\serverName\\H$").GetDirectories()`, the same error is thrown. This did not throw on _.net framework (4.6.1 at least)_. Empty root drive means there is no any directory or file, either hidden or system, like "System Volume Information" or "Recycle Bin".

[Related issue, but probably wrong place](https://github.com/dotnet/dotnet-api-docs/issues/3819)/

![image](https://user-images.githubusercontent.com/2863944/76411610-39dfaa80-63a3-11ea-982d-58a9e13b4533.png)

**Comment by carlossanlop** [comment]
I cannot repro with 3.1.

I had a spare USB drive that I formatted to NTFS, then FAT32, then ExFat. I ran your code and the "Success!" message was printed for all 3 cases.

@JeremyKuhne do you have an idea of what could be happening here?

**Comment by JeremyKuhne** [comment]
@carlossanlop deleting the hidden stuff is a pain as it keeps getting recreated. I can get it to repro if I blast the hidden folders (from an admin prompt) like follows:

```
L:\>takeown /f "System Volume Information" /r

SUCCESS: The file (or folder): "L:\System Volume Information" now owned by user "JEREMY-TOWER\jerem".

You do not have permissions to read the contents of directory "L:\System Volume Information".
Do you want to replace the directory permissions with permissions granting you
full control ("Y" for YES, "N" for NO or "C" to CANCEL)? y
SUCCESS: The file (or folder): "L:\System Volume Information\tracking.log" now owned by user "JEREMY-TOWER\jerem".

L:\>cacls "System Volume Information" /g jerem:F /t
Are you sure (Y/N)?y
processed dir: L:\System Volume Information
processed file: L:\System Volume Information\tracking.log

L:\>rd "System Volume Information" /s /q
```

I don't see the repro on .NET Framework.

**Comment by JeremyKuhne** [comment]
Note that they will come back after nuking- not sure what process does it specifically, but you have to work fast. :)

**Comment by carlossanlop** [comment]
Thanks, Jeremy.

It's my understanding that this folder should not be deleted or will cause unexpected problems on Windows, like the one we are seeing here. So the exception is expected, since we are not doing anything to mitigate its absence.

I'm wondering if we should prompt for a SVI folder right before `FindNextData` calls `GetData`, and if the folder does not exist, ask Windows to create it. But I'm not sure if this could have unintended consequences, or if mitigating the absence of the folder is something we should even do. Thoughts, @JeremyKuhne?

I found methods to retrieve Volume Information info, but I don't see one to create the SVI folder if does not exist:
https://docs.microsoft.com/en-us/windows/win32/fileio/volume-management-functions
https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getvolumeinformationa
https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getvolumeinformationbyhandlew

Only the Universal target platform offers methods that automatically create the SVI folder if it does not exist, but I assume we cannot call them here:
https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtlcreatesystemvolumeinformationfolder
https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtlcreatesystemvolumeinformationfolder

I was able to repro after forcing the deletion of the System Volume Information directory using the commands shared by Jeremy.

I completely forgot about the existence of the folder because my machine was not yet configured to show OS files. For the curious, this is how they are enabled:

> File Explorer window -> View tab -> Options -> Change folder and search options -> View tab ->
> Uncheck "Hide protected operating system files"

The exception is thrown by `GetData`, when trying to retrieve the first entry, which should be the root `H:\`:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs#L204

It makes a P/Invoke to `NtQueryDirectoryFile`, which gives us return status of `-1073741809`. Translated to an error code, it means 
 `ERROR_FILE_NOT_FOUND`:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Win32.cs#L54

**Comment by warappa** [comment]
Hello @carlossanlop ,
it seems to me that relying on at least one file system entry being found (i.e. SVI) is unexpected and unneccessary from an user's (read developer's) view.

Imagine a readonly media with no file system entry, which wasn't even created by Windows (therefore no SVI). Like empty CDs or USB sticks with a hardware readonly-switch or the forementioned empty (readonly) network share (Samba?). As a developer I should be able to query this media and find (without any exception) that it is indeed just empty.
Exception if drive not ready? OK. If no file system entry found? Not so.

**Comment by JeremyKuhne** [comment]
@carlossanlop as we had this working in .NET Framework (please try and validate that you can't repro there either) this may be a regression from my enumeration work. It is possible I missed some special "magic" that `FindFirstFile` does, but I don't see anything obvious. I also may have missed where we handled this particular situation in the original code.

**Comment by JeremyKuhne** [comment]
It looks like I did mess this up. 🙏🏻 Win32 does return an error, but we used to [swallow it](https://referencesource.microsoft.com/#mscorlib/system/io/filesystemenumerable.cs,278). Thanks for reporting this, @stukselbax!

**Comment by carlossanlop** [comment]
I have a fix for this and will submit a PR soon.

**Comment by stukselbax** [comment]
Will you port the fix once accepted to the next 3.1.* patch release? 3.1.3, probably?
